drafts/bitcartel-payment-disclosure/draft.rst

```diff
@@ -0,0 +1,336 @@
+
+The sender want sthe payment disclosure to be non-malleable, to prevent an attacker modifying details like the refund address.  To achieve this, the sender will sign the payment disclosure with the JoinSplitSig private key and append the signature to the end of the payment disclosure data.
+
+A new RPC call will be introduced to allow a third party to verify a payment disclosure.
+
+``z_validatepaymentdisclosure paymentdisclosure``
+
+Validates a payment disclosure and returns JSON output as follows:
+
+::
+
+    {
+    valid : true,
+    rawhex : "0011...ff",
+    version: 0
+    txid: "4e...",
+    jsindex: 0,
+    outputindex: 1,
+    paymentaddress: zABC123...,
+    value: 10.56100002
+    message: "pay me at z007",
+    signature: "10ab23..."
+    }
+
+Valid field is true if all the following conditions hold:
+
+- txid is confirmed in the blockchain
+- jsindex is valid for the txid
+- outputindex is within range [0..1]
+- paymentaddress is a valid shielded address for the network we are on
+- value is within range [0..MAX_MONEY]
+- message is within constraints e.g. number of characters, allowed characters
+- signature is valid for all of the above fields
+
+Otherwise an error field is returned explaining why the payment disclosure is invalid:
+
+::
+
+    {
+    valid : false,
+    ...
+    error : "The payment disclosuse is invalid because..."
+    }
+
+
+Implementation
+--------------
+TODO: Link to commits in Github
+
+A reference implementation will be added to zcashd as an experimental feature.  To enable payment disclosure, set the following two options to true:
+
+* `-paymentdisclosure=1`
+* `-experimentalfeatures=1`
+
+A third party trying to validate payment disclosure must track all transactions and thus enable the option:
+
+* `-txindex=1`
+
+In this implementation we will assume that once the feature is enabled, the node will create and store a payment disclosure for every transaction sent which involves a shielded address.
+
+When the sender calls RPC call ``z_sendmany`` and creates a shielded transaction in ``asyncrpcoperation.cpp``:
+
+#. retain the ephemeral joinSplitPrivKey used to sign the transaction
+#. retain the ephemeral secret key used for symmetric encryption of the note plaintext
+
+Record relevant data in a struct (or class) defined as follows:
+
+::
+
+    struct PaymentDisclosureKey {
+        uint256 txid            // primitives/transaction.h
+        size_t js;              // Index into CTransaction.vjoinsplit
+        uint8_t n;              // Index into JSDescription fields of length ZC_NUM_JS_OUTPUTS
+    };
+
+    struct PaymentDisclosureInfo {
+        uint8_t version;          // 0 = experimental, 1 = first production version, etc.
+        uint256 esk;              // zcash/NoteEncryption.cpp
+        uint256 joinSplitPrivKey; // primitives/transaction.h
```

Comment:
Instead of uint256, may be more convenient to store full 64 byte output from libsodium crypto_sign_keypair() rather than extracting the private key and later having to reconstruct output via tx.joinSplitPubKey.  The full 64 bytes is required by crypto_sign_detached().